### PR TITLE
Replace the json.org library with Open JSON

### LINKED ISSUE #2672: Please do not depend on org.json:json which is non-free software
Please do not depend on org:json:json which is non-free software. It contains the infamous license clause "The Software shall be used for Good, not Evil." This dependency could be replaced by existing free software.

https://wiki.debian.org/qa.debian.org/jsonevil

### PR DESCRIPTION
**Pull request by ssoloff**
Resolves #2672.

The json.org library uses a license that is incompatible with the GPL.  Open JSON uses the Apache License 2.0 which is compatible with GPLv3 (but not GPLv2).  We use other libraries licensed under the Apache License 2.0 (e.g. Guava), so this doesn't introduce any new issues.  However, we should consider opening a separate issue to discuss changing the TripleA license to GPLv3.

Open JSON is pretty much a drop-in replacement for the json.org library.  However, it is based on an older API version and is missing some of the newer convenience methods we rely on.  I added equivalent methods in the new `OpenJsonUtils` class to provide this missing functionality.  These can be easily removed if a new version of Open JSON provides equivalent functionality in the future.

In summary, the following replacements were made:

Old method | New method
:-- | :--
`JSONArray#spliterator()` | `OpenJsonUtils#stream(JSONArray)`
`JSONObject#optEnum()` | `OpenJsonUtils#optEnum(JSONObject, ...)`
`JSONObject#toMap()` | `OpenJsonUtils#toMap(JSONObject)`

#### Testing

Unit tests were added for all new methods in `OpenJsonUtils`.

I smoke tested the following three features that rely on parsing JSON documents:

* Download maps
* Lobby
* PBF via `forums.triplea-game.org`

### CODE REVIEW

File: src/main/java/games/strategy/util/OpenJsonUtils.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+  public static Stream<Object> stream(final JSONArray jsonArray) {
+    checkNotNull(jsonArray);
+
+    final int size = jsonArray.length();
+    final List<Object> list = new ArrayList<>(size);
+    for (int i = 0; i < size; ++i) {
+      list.add(jsonArray.get(i));
+    }
+    return list.stream();
+  }
```
Comment (RoiEXLab) [inline src/main/java/games/strategy/util/OpenJsonUtils.java]: I'd suggest using a [StreamBuilder](https://docs.oracle.com/javase/9/docs/api/java/util/stream/Stream.Builder.html) instead.
The javadocs explicitly state:
> A mutable builder for a Stream. This allows the creation of a Stream by generating elements individually and adding them to the Builder (without the copying overhead that comes from using an ArrayList as a temporary buffer.)

File: src/main/java/games/strategy/util/OpenJsonUtils.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+   *
+   * @return A list of the specified JSON array's elements converted to their equivalent Java types.
+   */
+  public static List<Object> toList(final JSONArray jsonArray) {
+    checkNotNull(jsonArray);
+
+    final int size = jsonArray.length();
+    final List<Object> list = new ArrayList<>(size);
+    for (int i = 0; i < size; ++i) {
+      list.add(unwrap(jsonArray.get(i)));
```
Comment (RoiEXLab) [inline src/main/java/games/strategy/util/OpenJsonUtils.java]: You could make use of the stream method you created above here 😄 
```java
return stream(jsonArray).map(OpenJsonUtils::unwrap).collect(Collectors.toList());
```

File: src/main/java/games/strategy/util/OpenJsonUtils.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+   *
+   * @return A map of the specified JSON object's properties converted to their equivalent Java types. The keys are the
+   *         property names; the values are the property values.
+   */
+  public static Map<String, Object> toMap(final JSONObject jsonObject) {
+    checkNotNull(jsonObject);
+
+    final Map<String, Object> map = new HashMap<>(jsonObject.length());
+    for (final String name : jsonObject.keySet()) {
+      map.put(name, unwrap(jsonObject.get(name)));
```
Comment (RoiEXLab) [inline src/main/java/games/strategy/util/OpenJsonUtils.java]: Streaming works here as well, although not that pretty ^^
```java
return jsonObject.keySet().stream()
    .collect(Collectors.toMap(
        Function.indentity(),
        name -> unwrap(jsonObject.get(name))));
```

**Comment by RoiEXLab** [comment]
@ssoloff 
About the License Problem:
Also I found an old post of mine about the GPL, because I recalled bringing up this topic once:
https://gitter.im/triplea-game?at=5752b15c813c217c21da186e
(It's a private gitter room)
@DanVanAtta Told me the approval of all original authors is required for this, which is going to be problematic, since @veqryn is probably the only inactive author we are able to "talk" with.
I believe commons-math also uses the apache license btw.
But I'm not sure if this is actually Problematic:
> Each version is given a distinguishing version number.  If the Program
specifies a version number of this License which applies to it and "any
later version", you have the option of following the terms and conditions
either of that version or of ***any later version*** published by the Free
Software Foundation.  If the Program does not specify a version number of
this License, you may choose any version ever published by the Free Software
Foundation.

If I understand the legal text correctly, we are free to use TripleA under the terms of the GPL v3, but correct me if I'm wrong, I'm not a lawyer of any means, nor is english my native language

File: src/main/java/games/strategy/util/OpenJsonUtils.java (hunk 3 above)
Comment (RoiEXLab) [reply src/main/java/games/strategy/util/OpenJsonUtils.java]: Even better:
```java
final Function<String, Object> keyToValue = jsonObject::get;
return jsonObject.keySet().stream()
    .collect(Collectors.toMap(
        Function.indentity(),
        keyToValue.andThen(OpenJsonUtils::unwrap)));
// Alternatively
return jsonObject.keySet().stream()
    .collect(Collectors.toMap(
        Function.indentity(),
        ((Function<String, Object>)jsonObject::get)).andThen(OpenJsonUtils::unwrap)));
```

File: src/main/java/games/strategy/util/OpenJsonUtils.java (hunk 1 above)
Comment (ssoloff) [reply src/main/java/games/strategy/util/OpenJsonUtils.java]: Thanks for the reference.  I'll look into making that change.

File: src/main/java/games/strategy/util/OpenJsonUtils.java (hunk 2 above)
Comment (ssoloff) [reply src/main/java/games/strategy/util/OpenJsonUtils.java]: Derp, of course. :smile:

**Comment by ssoloff** [comment]
@RoiEXLab 

> About the License Problem

Thanks for the reference.  And, yes, there's probably more to it than simply saying we're now using GPLv3. :smile:

> If I understand the legal text correctly, we are free to use TripleA under the terms of the GPL v3

I'm not a lawyer either, but I would interpret it with one minor difference: a TripleA _user_ has the option to follow the terms of the GPLv2 or a later version.  We (the TripleA developers) are providing it under the GPLv2, which I believe means we should ensure a user _could_ choose to use the terms of the GPLv2.  But if we include software that is not compatible with the GPLv2, that would seem to prevent users from choosing that licensing model.

BTW, here's another good [discussion](https://opensource.stackexchange.com/questions/1357/can-i-link-a-apache-2-0-library-into-software-under-gplv2) I found on this issue.

File: src/main/java/games/strategy/util/OpenJsonUtils.java (hunk 3 above)
Comment (ssoloff) [reply src/main/java/games/strategy/util/OpenJsonUtils.java]: Unfortunately, `Collectors#toMap()` calls `Map#merge()`, which doesn't seem to like `null` values (which are legal based on the json.org library's API):

```
java.lang.NullPointerException
	at java.util.HashMap.merge(HashMap.java:1224)
	at java.util.stream.Collectors.lambda$toMap$58(Collectors.java:1320)
	at java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169)
	at java.util.Iterator.forEachRemaining(Iterator.java:116)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
	at games.strategy.util.OpenJsonUtils.toMap(OpenJsonUtils.java:116)
	at games.strategy.util.OpenJsonUtilsTest.toMap_ShouldConvertNullSentinelValueToNull(OpenJsonUtilsTest.java:174)
	<...snip...>
```

Comment (RoiEXLab) [reply src/main/java/games/strategy/util/OpenJsonUtils.java]: Nevermind then
